gson/src/test/java/com/google/gson/GsonTest.java

```diff
@@ -368,4 +360,223 @@ public CustomClass3() {
+          if (type.getRawType() == CustomClassA.class) {
+            // Factory will throw for CustomClassB1; discard exception
+            try {
+              gson.getAdapter(CustomClassB1.class);
+              fail("Expected exception");
+            } catch (Exception e) {
+              assertEquals("test exception", e.getMessage());
+            }
+
+            @SuppressWarnings("unchecked")
+            TypeAdapter<T> adapter = (TypeAdapter<T>) alternativeAdapter;
+            return adapter;
+          }
+          else if (type.getRawType() == CustomClassB1.class) {
+            gson.getAdapter(CustomClassC.class);
+            // Will throw exception
+            gson.getAdapter(CustomClassX.class);
+
+            throw new AssertionError("Factory should have thrown exception for CustomClassX");
+          }
+          else if (type.getRawType() == CustomClassC.class) {
+            // Will return future adapter due to cyclic dependency B1 -> C -> B1
+            TypeAdapter<?> adapter = gson.getAdapter(CustomClassB1.class);
+            assertTrue(adapter instanceof FutureTypeAdapter);
+            // Pretend this factory somehow leaks this FutureTypeAdapter
+            leakedAdapter.set(adapter);
+            return new DummyAdapter<T>();
+          }
+          else if (type.getRawType() == CustomClassX.class) {
+            // Always throw exception
+            throw new RuntimeException("test exception");
+          }
+
+          throw new AssertionError("Requested adapter for unexpected type: " + type);
+        }
+      })
+      .create();
+
+    assertSame(alternativeAdapter, gson.getAdapter(CustomClassA.class));
+    // Gson must not have cached broken adapters for CustomClassB1 and CustomClassC
+    try {
+      gson.getAdapter(CustomClassB1.class);
+      fail("Expected exception");
+    } catch (Exception e) {
+      assertEquals("test exception", e.getMessage());
+    }
+    try {
+      gson.getAdapter(CustomClassC.class);
+      fail("Expected exception");
+    } catch (Exception e) {
+      assertEquals("test exception", e.getMessage());
+    }
+
+    // Leaked adapter should have been marked as "broken"
+    try {
+      leakedAdapter.get().fromJson("{}");
+      fail("Expected exception");
+    } catch (IllegalStateException e) {
+      assertEquals("Broken adapter has been leaked by TypeAdapterFactory", e.getMessage());
+    }
+  }
+
+  /**
+   * Verifies that two threads calling {@link Gson#getAdapter(TypeToken)} do not see the
+   * same unresolved {@link FutureTypeAdapter} instance, since that would not be thread-safe.
+   *
+   * This test constructs the cyclic dependency CustomClassA -> CustomClassB1 -> CustomClassA
+   * and lets one thread wait after the adapter for CustomClassB1 has been obtained (which still
+   * contains the nested unresolved FutureTypeAdapter for CustomClassA).
+   */
+  public void testGetAdapterFutureAdapterConcurrency() throws Exception {
```

Comment:
This test is quite verbose but it should hopefully cover exactly the situation #625 is about; reverting the `Gson` changes of this pull request but keeping these `GsonTest` changes triggers the bug described in that issue (`IllegalStateException` being thrown).